In [21]:
!pip install tabulate

In [25]:
!pip install scikit-learn


     ---------------------------------------- 0.0/9.2 MB ? eta -:--:--
     ---------------------------------------- 0.1/9.2 MB 3.2 MB/s eta 0:00:03
     - -------------------------------------- 0.3/9.2 MB 3.5 MB/s eta 0:00:03
     - -------------------------------------- 0.3/9.2 MB 2.7 MB/s eta 0:00:04
     - -------------------------------------- 0.4/9.2 MB 2.4 MB/s eta 0:00:04
     -- ------------------------------------- 0.5/9.2 MB 2.2 MB/s eta 0:00:04
     --- ------------------------------------ 0.7/9.2 MB 2.7 MB/s eta 0:00:04
     ---- ----------------------------------- 0.9/9.2 MB 2.9 MB/s eta 0:00:03
     ---- ----------------------------------- 1.0/9.2 MB 2.7 MB/s eta 0:00:04
     ----- ---------------------------------- 1.2/9.2 MB 3.0 MB/s eta 0:00:03
     ------ --------------------------------- 1.4/9.2 MB 3.2 MB/s eta 0:00:03
     ------- -------------------------------- 1.6/9.2 MB 3.3 MB/s eta 0:00:03
     -------- ------------------------------- 1.9/9.2 MB 3.4 MB/s eta 0

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Exploration & Compréhension

In [2]:
data = pd.read_csv("jobs.csv", encoding="latin-1")
data.head()


Company                                    Job Title   
0              SGS                        Clinical Data Analyst  \
1          Ocorian                       AML/CFT & Data Analyst   
2           Cricut                    Machine Learning Engineer   
3      Bosch Group         Application Developer & Data Analyst   
4  Publicis Groupe  Data Engineer Full time (Public Sector) USA   

                          Location   Job Type Experience level   Salary   
0    Richardson, TX, United States  Full Time      Entry-level   48K+ *  \
1                 Ebène, Mauritius  Full Time      Entry-level   48K+ *   
2  South Jordan, UT, United States  Full Time              NaN   90K+ *   
3                 Nonantola, Italy  Full Time      Entry-level   48K+ *   
4     Arlington, VA, United States  Full Time        Mid-level    108K+   

                          Requirment of the company    
0  Computer Science,Data quality,Genetics,Mathema...  \
1           Agile,Data management,Finance,Security,,   
2  Agile,Architecture,AWS,Computer Science,Comput...   
3       Engineering,Industrial,Oracle,Power BI,R,R&D   
4  AWS,Azure,Computer Science,Consulting,Dataflow...   

                                          Facilities  
0                                               ,,,,  
1                                               ,,,,  
2                             Career development,,,,  
3                                               ,,,,  
4  Flex hours,Flex vacation,Parental leave,Unlimi...

In [3]:
data.shape

(3198, 8)

Nous avons un dataset de 3198 entrées et 8 champs

In [4]:
data.dtypes

Company                       object
Job Title                     object
Location                      object
Job Type                      object
Experience level              object
Salary                        object
Requirment of the company     object
Facilities                    object
dtype: object

Toutes les variables sont des objets. Bon à savoir !

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3198 entries, 0 to 3197
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Company                     3197 non-null   object
 1   Job Title                   3197 non-null   object
 2   Location                    3197 non-null   object
 3   Job Type                    3197 non-null   object
 4   Experience level            2962 non-null   object
 5   Salary                      3009 non-null   object
 6   Requirment of the company   3198 non-null   object
 7   Facilities                  3198 non-null   object
dtypes: object(8)
memory usage: 200.0+ KB


- Compréhension & Analyse: Nous avons 3198 entrées. 
    - Seules les colonnes "Facilities" et "Requirment of company" sont toutes bien remplies
    - Les colonnes "Company", "Job Title", "Location" ont seulement une valeur manquante
    - La colonne "Experience level" a 2962 entrées sur 3198. Donc près de 93% de valeurs remplies et 7% de valeurs manquantes 
    - La colonne "Salary" a 3009 entrées sur 3198. Donc 94% de valeurs remplies et 6% de valeurs manquantes
    

In [6]:
# Afficher un résumé statistique du DataFrame
summary = data.describe(include='all')
summary


Company      Job Title          Location   Job Type   
count              3197           3197              3197       3197  \
unique             1106           2138              1117          3   
top     Publicis Groupe  Data Engineer  Bengaluru, India  Full Time   
freq                126            105                90       3116   

       Experience level    Salary   
count              2962      3009  \
unique                4       218   
top        Senior-level   115K+ *   
freq               1876       253   

                               Requirment of the company  Facilities  
count                                                3198       3198  
unique                                               2600        777  
top     Big Data,Business Intelligence,Data analysis,E...       ,,,,  
freq                                                   12        542

- "count": Cela indique le nombre total de valeurs non manquantes (non nulles) dans chaque colonne du DataFrame. Par exemple, il y a 3197 valeurs non manquantes pour la colonne "Company", "Job Title", "Location", etc.

- "unique": Cela représente le nombre d'éléments uniques dans chaque colonne du DataFrame. Par exemple, il y a 1106 entreprises uniques, 2138 titres de poste uniques, 1117 emplacements uniques, etc.

- "top": Cela indique la valeur la plus fréquente (top) dans chaque colonne. Par exemple, "Publicis Groupe" est l'entreprise la plus fréquente dans la colonne "Company", "Data Engineer" est le titre de poste le plus fréquent, "Bengaluru, India" est le lieu le plus fréquent, etc.

- "freq": Cela représente la fréquence (nombre d'occurrences) de la valeur la plus fréquente (top) dans chaque colonne. Par exemple, "Publicis Groupe" apparaît 126 fois dans la colonne "Company", "Data Engineer" apparaît 105 fois dans la colonne "Job Title", "Bengaluru, India" apparaît 90 fois dans la colonne "Location", etc.


Une remarque: Nous n'avons pas de valeurs numériques !

# Data Cleaning

Cette étape consiste à nettoyer les données en corrigeant les erreurs, en supprimant les impuretés pour ne conserver que les valeurs propres, correctes, utiles et pertinentes. Nous serons amenés à:
- Gérer les doublons
- Gérer les valeurs manquantes
- Gérer les valeurs abérrantes et atypiques

### Gestion des doublons

In [7]:
data.duplicated().sum()

202

Nous avons 202 doublons. Nous allons les supprimer pour éviter les conflits de données.

In [8]:
data.drop_duplicates(inplace = True)
data.duplicated().sum()

0

Les doublons ont bien été supprimés

### Gestion des valeurs manquantes

In [9]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2996 entries, 0 to 3196
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Company                     2995 non-null   object
 1   Job Title                   2995 non-null   object
 2   Location                    2995 non-null   object
 3   Job Type                    2995 non-null   object
 4   Experience level            2768 non-null   object
 5   Salary                      2824 non-null   object
 6   Requirment of the company   2996 non-null   object
 7   Facilities                  2996 non-null   object
dtypes: object(8)
memory usage: 210.7+ KB


Nous avons à présent 2996 entrées.
- Les colonnes "Company", "Job Title", "Location", "Job Type" ont 2995 entrées donc ont 01 valeur manquante.
- La colonne "Experience level" possède 2768 entrées donc elle possède 228 valeurs manquantes soit environ 7,6% 
- La colonne "Salary" a 2824 entrées. Elle a donc 172 valeurs manquantes soit 5,7% 

### Colonne Company

In [10]:
missing_values_count = data["Company"].isna().sum()
print("Number of missing values in 'Company' column:", missing_values_count)


Number of missing values in 'Company' column: 1


In [11]:
data.loc[data["Company"].isna()]


Company Job Title Location Job Type Experience level Salary   
797     NaN       NaN      NaN      NaN              NaN    NaN  \

    Requirment of the company  Facilities  
797                      ,,,,,       ,,,,

### La ligne n'a aucune information qui soit utile. Nous allons donc la supprimer

In [12]:
# Supprimer les lignes où "Company" est NaN et mettre à jour le DataFrame data
data.dropna(subset=["Company"], inplace=True)

# Vérification
data.loc[data["Company"].isna()]


Empty DataFrame
Columns: [Company, Job Title, Location, Job Type, Experience level, Salary, Requirment of the company , Facilities]
Index: []

La ligne ligne a bien été supprimée. Passons à la colonne Job Title

### Colonne Job Title

In [13]:
data["Job Title"].isna().sum()

0

### Colonne Location 

In [14]:
data["Location"].isna().sum()

0

### Colonne Job Type

In [15]:
data["Job Type"].isna().sum()

0

Les valeurs manquantes des colonnes Job Type, Location,Job Title et Company étaient dans la meme entrée et la ligne est désormais supprimée.
- Passons à la colonne Experience level

### Experience level 

In [16]:
data["Experience level"].isna().sum()

227

Il y a exactement 227 valeurs manquantes. Allons-nous les supprimer ? Nous allons éviter au maximum de supprimer. Nous allons tenter une amputation. 

In [17]:
data["Experience level"]


0        Entry-level
1        Entry-level
2                NaN
3        Entry-level
4          Mid-level
            ...     
3190       Mid-level
3191       Mid-level
3192    Senior-level
3193     Entry-level
3196     Entry-level
Name: Experience level, Length: 2995, dtype: object

In [18]:
# Obtenir le nombre d'occurrences de chaque valeur unique dans la colonne "Experience level"
experience_counts = data["Experience level"].value_counts()

# Obtenir la valeur la plus fréquente dans la colonne "Experience level" (en cas d'égalité, on prend la première)
most_frequent_experience = experience_counts.idxmax()
most_frequent_experience
# # Remplacer les valeurs NaN par la valeur la plus fréquente dans la colonne "Experience level"
# data["Experience level"] = data["Experience level"].fillna(most_frequent_experience)

# # Afficher le nombre d'occurrences de chaque valeur unique dans la colonne "Experience level"
# print("Nombre d'occurrences de chaque valeur unique dans la colonne 'Experience level':")
# print(experience_counts)

# # Afficher le DataFrame mis à jour avec les valeurs NaN remplacées
# print("\nDataFrame mis à jour avec les valeurs NaN remplacées:")
# print(data["Experience level"])


'Senior-level'

In [19]:
xp_level = data.loc[data["Experience level"].isna()]
xp_level

Company                            Job Title   
2                       Cricut            Machine Learning Engineer  \
11                   NielsenIQ     Data Analyst - Revenue Optimizer   
18                       Issuu                           BI Analyst   
36                 Bosch Group           Data Science Lead (Hybrid)   
69                      SSENSE    Data Analyst, Customer Experience   
...                        ...                                  ...   
3149            Syngenta Group  Agronomy Data Scientist (M/W) - CDI   
3157    Precision for Medicine                        Data Engineer   
3163        Informa Group Plc.             Data Operations Engineer   
3172  Precision Medicine Group                        Data Engineer   
3177    MUFG Investor Services           Data Engineer (Contractor)   

                               Location   Job Type Experience level   Salary   
2       South Jordan, UT, United States  Full Time              NaN   90K+ *  \
11                  Toronto, ON, Canada  Full Time              NaN   80K+ *   
18                                Braga  Full Time              NaN   48K+ *   
36    Mount Prospect, IL, United States  Full Time              NaN   63K+ *   
69                 Montreal, QC, Canada  Full Time              NaN   80K+ *   
...                                 ...        ...              ...      ...   
3149              Saint-Sauveur, France  Full Time              NaN   90K+ *   
3157              Remote, United States  Full Time              NaN     70K+   
3163                Toronto, ON, Canada  Full Time              NaN   60K+ *   
3172              Remote, United States  Full Time              NaN     70K+   
3177                   Limassol, Cyprus  Full Time              NaN   95K+ *   

                             Requirment of the company    
2     Agile,Architecture,AWS,Computer Science,Comput...  \
11    Business Analytics,Business Intelligence,Data ...   
18    Business Analytics,Business Intelligence,Data ...   
36    A/B testing,Data visualization,Economics,Engin...   
69    Business Intelligence,Computer Science,CX,Data...   
...                                                 ...   
3149  Computer Science,D3,Data analysis,Data Mining,...   
3157        AWS,Biology,Data management,Git,Linux,MySQL   
3163        Agile,AWS,DataOps,Engineering,ETL,Pipelines   
3172        AWS,Biology,Data management,Git,Linux,MySQL   
3177     Agile,Computer Science,Engineering,ETL,Python,   

                                             Facilities  
2                                Career development,,,,  
11            Career development,Startup environment,,,  
18        Competitive pay,Equity,Health care,Insurance,  
36     Career development,Competitive pay,Health care,,  
69            Career development,Startup environment,,,  
...                                                 ...  
3149                             Career development,,,,  
3157  Health care,Insurance,Parental leave,Salary bo...  
3163  Career development,Flex hours,Flex vacation,He...  
3172  Health care,Insurance,Parental leave,Salary bo...  
3177                                Competitive pay,,,,  

[227 rows x 8 columns]

In [20]:
import pandas as pd

# Spécifiez l'encodage approprié (par exemple, "latin-1")
data = pd.read_csv("jobs.csv", encoding="latin-1")

# Remplacer les valeurs manquantes par "No Experience"
data["Experience level"].fillna("No Experience", inplace=True)

# Afficher le DataFrame mis à jour
print(data["Experience level"].isnull().sum())


0


### Colonne Salary

In [21]:
data["Salary"].isnull().sum()

189

Nous avons 189 entrées manquantes dans la colonne Salary. Nous allons tenter une imputation. Pour le moment, nous allons remplacer les valeurs manquantes par "No Salary"

In [22]:
data["Salary"].fillna("No Salary", inplace = True)
print(data["Salary"].isnull().sum())

0


Il n'existe plus de valeurs manquantes 

#### Gestion de valeurs manquantes terminées !

## Gestion des valeurs Abérrantes